---
title: "RTIGER"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{RTIGER}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  echo = TRUE
)
```

```{r setup}
library(RTIGER)
```
\

## Introduction
Accurate identification of meiotic crossing-over sites (COs) is essential for correct genotyping of recombining samples. RTIGER is a method for predicting genome-wide COs using allele-counts at pre-defined SNP marker positions. RTIGER trains a Hidden Markov Model (HMM) where genomic states (homozygous parent_1, homozygous parent_2 or heterozygous) correspond to the hidden state and the allele-counts as the observed variable. COs are identified as transitions in the HMM state.

To account for variation in the coverage of sequencing data, RTIGER uses Viterbi Path Algorithm and the `rigidity` parameter. This parameter defines the minimum number of SNP markers required to support a state-transition. This filters out low-confidence state-transitions, improving COs identification performance.
\

<!-- ################################################################################ -->
## Installation
#### Pre-Requisites:
* R: Version > X_X_X
* Julia-1.0.5 (Which versions of Julia can be supported?): Julia needs to be installed and available in the environment^[https://www.geeksforgeeks.org/how-to-setup-julia-path-to-environment-variable/?ref=lbp]
* REQUIRED R LIBRARIES WOULD BE INSTALLED AUTOMATICALLY... RIGHT?

\

<!-- ################################################################################ -->
### Preparing input data:
RTIGER uses the allele-count information at the SNP marker positions. The SNP markers correspond to differences between the two genotypes (i.e. parent_1 vs parent_2). RTIGER requires as input one allele-count file for each sample. The allele-count file should be in tab-separated value format, where each row corresponds to a SNP marker. The format of the file is described below:

```{r echo=FALSE, results='asis'}
library(knitr)
Column= 1:6
Field = c("SeqID", "Pos", "RefA", "RefC", "AltA", "AltF")
Type = c("string", "int (>=0)", "char", "int (>=0)", "char", "int (>=0)")
Description = c("Chromosome ID", "Position of the SNP marker", "Reference allele", "Number of reads with reference allele", "Alternate allele", "Number of reads with alternate allele")

kable(data.frame(Column, Field, Type, Description), align = "l", caption='File Format for allele frequency file', justify='centred', width='minimum')
```

The SNPs can be identified using any generic SNP identification pipeline^[For example: https://www.ebi.ac.uk/sites/ebi.ac.uk/files/content.ebi.ac.uk/materials/2014/140217_AgriOmics/dan_bolser_snp_calling.pdf].


SNPs in repetitive regions should be filtered out. Further, as crossing-over usually takes place in syntenic regions between the two genome, for best results, only SNPs in syntenic regions should be selected as markers. If whole genome assemblies are present for both genomes, then this can be easily achieved using methods like SyRI^[https://genomebiology.biomedcentral.com/articles/10.1186/s13059-019-1911-0].

**NOTE 1**: RTIGER assumes that all samples have similar sequencing coverage and, hence, similar distribution of the allele-count values. It does not check or normalise for sequencing coverage variation.

**NOTE 2**: Crossing-over resolution depends on sequenced marker density. Low sequencing coverage could result in few informative markers, which in turn could decrease resolution CO prediction.

**NOTE 3**: RTIGER is designed to be robust against individual outliers, however, the user should check for "bad" markers, i.e. marker positions that are prone to mismapping. These markers result in high allele-count at that position.<!--  as well as in a high, artificial agreement in this position across samples. -->
    
\

<!-- ################################################################################ -->
## Using RTIGER

#### Setting up Julia environment:
RTIGER uses Julia to perform computationally intensive model training. All Julia packages that are used by RTIGER can be installed using using:
```{r echo= TRUE, eval=FALSE}
library(RTIGER)
setupJulia()
```
This step is **necessary** when using RTIGER for the first time, but can be skipped for later analysis as all required Julia packages would already be installed.

The Julia functions need to be loaded in the R environment using:
```{r echo=TRUE, eval=FALSE}
sourceJulia()
```
This step is required everytime when using RTIGER.

\

#### Creating input objects
The primary input for RTIGER is a data-frame termed `expDesign`. The first column of `expDesign` should have paths to allele-count files for all samples and the second column should have unique samples IDs.

```{r echo = TRUE}
# Get paths to example allele count files originating from a
# cross between Col-0 and Ler accession of the A.thaliana
file_paths = list.files(system.file("extdata",  package = "RTIGER"), full.names = TRUE)

# Get sample names
sampleIDs <- basename(file_paths)

# Create the expDesign object
expDesign = data.frame(files=file_paths, name=sampleIDs)

print(expDesign)
```

RTIGER also requires chromosome lengths for the parent_1. These need to be provided as a named vector where the values are chromosome lengths and the names are chromosome ids.

``` {r echo=TRUE}
# Get chromosome lengths for the example data included in the package
chr_len <- RTIGER::ATseqlengths
names(chr_len) <- c('Chr1' , 'Chr2', 'Chr3', 'Chr4', 'Chr5')
print(chr_len)
```

\

#### Finding crossing-over sites using RTIGER
RTIGER does model training, COs identification, per sample and summary plots creation using a single function.
```{r echo=TRUE, eval=FALSE}
myres = RTIGER(expDesign = expDesign,
               outputdir = "/srv/netscratch/dep_mercier/grp_schneeberger/projects/SynSearch/tests",
               seqlengths = chr_len,
               rigidity = 200)
```

The `rigidity` parameter defines the required minimum number of continuous markers that together support a state change of the HMM model. Smaller `rigidity` values increase the sensitivity in detecting COs that are close to each other, but may result in false-positive CO identification because of variation in sequencing coverage. Larger `rigidity` values improve precision but COs that are close to each other might not be identified. **Users are supposed to test and adjust `rigidity` based on their specific experimental setup**.
\
<!-- ################################################################################ -->

## RTIGER Output:
RTIGER identifies COs for each sample level and provides summary plots and statistics for each sample as well as for the entire population.

\

#### Per sample output
RTIGER creates a folder for each sample in the `outputdir`. This folder contains:

* `GenotypePlot.pdf`: Graphical representation of the allele-counts, allele-count ratio, and genotypes
* `GenotypeBreaks.bed`: BED file providing genomic regions corresponding to different genotypes
* `P1/P2/Het.bed`: BED files containing the markers present in genomic regions having genotype: homozygous parent 1, homozygous parent 2, or heterozygous, respectively
* `P1/P2.bw`: BigWig file containing the number of reads per marker position supporting parent 1 and parent 2, respectively
* `CountRatio.bw`: BigWig file containing the ratio of number of reads supporting parent 1 to number of reads supporting number 2 at the marker positions

#### Summary plots for the population (THE PLOTS AND THE DESCRIPTION HERE NEEDS MORE WORK)
RTIGER creates four summary plots after aggregating results for all samples. 

* `COs-per-Chromosome.pdf`: Distribution of number of cross-overs per chromosome
* `CO-count-perSample.pdf`: Number of cross-overs in each sample
* `Goodness-Of-fit.pdf`: ****NEED DESCRIPTION FROM RAFA****
* `GenomicFrequencies.pdf`: Distribution of cross-overs along the length of chromosomes


### Analysing backcrossed populations
Backcrossed populations are formed by crossing a hybrid organism with one of its parent. These populations are different from the populations based on outcrossing as only two genomic states are possible (homozygous for the backrossed parent and heterozygous for both parents). To identify COs in such population, set `nstates=2` in the RTIGER command.
```{r echo=TRUE, eval=FALSE}
myres = RTIGER(expDesign = expDesign, 
               outputdir = "PATH/TO/OUTPUT/DIR",
               seqlengths = chr_len,
               rigidity = 200, 
               nstates=2)
```


<!-- myres = RTIGER(expDesign = expDesign, outputdir = paste0("/srv/netscratch/dep_mercier/grp_schneeberger/projects/rtiger/test/"), seqlengths =chr_len, rigidity = 200) -->
    


<!-- ## Running RTIGER: -->
<!--     ## Examples of using other running modes for RTIGER (changing parameters)? -->
<!--     ## Generating CO-Lists from rtiger_out object (if not generated automatically) -->
<!--     ## Description of using other ease-of-life features provided by RTIGER     -->
    

################################################################################
## Cite:
    Citation        

################################################################################
## Appendix:

### Effect of varying `rigidity (R)` on crossover identification:
<!-- Do we have estimates on how RTIGER performance varies with number of samples? Markers? Is it possible to use it for 1 sample with only 100 markers (rows in allele_count file)? Might be important to describe this here. -->

        
################################################################################
## Session info
```{r echo=FALSE, eval=TRUE}
sessionInfo()
```

